src/game.cpp

```diff
@@ -10242,245 +10242,226 @@ void game::grab()
+        }
+		// From infinite source to the ground somewhere else. The target has
+		// infinite space and the liquid can not be used from there anyway.
+        if( !has_keg && is_infinite && source_pos != nullptr ) {
+            add_msg( m_info, _( "Clearing out the %s would take forever." ), m.name( *source_pos ).c_str() );
+            return;
+        }
 
-            if( cont->ammo_remaining() && cont->ammo_current() != liquid.typeId() ) {
-                add_msg( m_info, _( "You can't mix loads in your %s." ), cont->tname().c_str() );
-                return -1;
-            }
+        // One can either pour into a keg *or* on the ground. It is never possible to do both on
+        // the same square because kegs have the NOITEM flag. In other words:
+        // assert( !( has_keg && can_put ) );
+        // This allows to start the activity with only the target position as parameter.
+        if( create_activity() ) {
+            serialize_liquid_target( u.activity, target_pos );
+            return;
+        }
 
-            add_msg( _(" You pour %1$s into the %2$s." ), liquid.tname().c_str(), cont->tname().c_str() );
-            auto qty = std::min( liquid.charges, cont->ammo_capacity() - cont->ammo_remaining() );
-            liquid.charges -= qty;
-            cont->ammo_set( liquid.typeId(), cont->ammo_remaining() + qty );
-            if( liquid.charges > 0 ) {
-                add_msg( _( "There's some left over!" ) );
-                return liquid.charges;
+        if( has_keg ) {
+            if( !iexamine::pour_into_keg( target_pos, liquid ) ) {
+                return;
             }
-            return 0;
-
         } else {
-            item tmp_liquid = liquid;
-            std::string err;
-            bool allow_bucket = cont == &u.weapon || !u.has_item( *cont );
-            if( !cont->fill_with( tmp_liquid, err, allow_bucket ) ) {
-                add_msg( m_info, "%s", err.c_str() );
-                return -1;
-            }
-            u.inv.unsort();
-            if( tmp_liquid.charges == 0 ) {
-                add_msg(_("You pour %1$s into your %2$s."), liquid.tname().c_str(), cont->type_name().c_str());
-            } else {
-                add_msg(_("You fill your %1$s with some of the %2$s."), cont->type_name().c_str(), liquid.tname().c_str());
-                add_msg(_("There's some left over!"));
-            }
-            return tmp_liquid.charges;
+            m.add_item_or_charges( target_pos, liquid, 1 );
+            liquid.charges = 0;
         }
-        return -1;
+
+        u.mod_moves( -100 );
+    } );
+    if( liquid.rotten() ) {
+        // Pre-select this one as it is the most likely one for rotten liquids
+        menu.selected = menu.entries.size() - 1;
     }
-    return -1;
+
+    if( menu.entries.empty() ) {
+        return false;
+    }
+    // Has no associated action, it's there to show that cancelling is an option.
+    menu.addentry( -1, true, MENU_AUTOASSIGN, _( "Cancel" ) );
```

Comment:
Setting `menu.return_invalid = true` would allow escaping from the menu which is more consistent. It also stops the hotkey `1` being inconsistently applied to the cancel or vehicle options.